### PR TITLE
2.x: add ConsumableX to the base types, update method signatures

### PR DESCRIPTION
**Pull request by akarnokd**
This PR factors out the `XSubscriber` types, makes sure the base types implement `ConsumableX`
and methods accepting X in some way now accept `ConsumableX` type.

Wrapping into the base types are not yet eliminated.

### CODE REVIEW

File: src/main/java/io/reactivex/Completable.java [hunk 1]
```diff
@@ -559,7 +558,7 @@ public void onSubscribe(Subscription s) {
      * @throws NullPointerException if flowable is null
      */
     @SchedulerSupport(SchedulerKind.NONE)
-    public static <T> Completable fromNbpObservable(final Observable<T> observable) {
+    public static <T> Completable fromObservable(final Observable<T> observable) {
```
Comment (stealthcode) [inline src/main/java/io/reactivex/Completable.java]: Should these kinds of `fromX(X)` methods change to something like `fromX(ConsmuableX)`? I think its sufficient to adopt the more generic api but still call out to `Observable` in the name and in the javadoc comment.

Comment (akarnokd) [inline src/main/java/io/reactivex/Completable.java]: Yes, I missed a few places. Updated the PR.

**Comment by akarnokd** [comment]
PR updated with the few places where I forgot to replace with the nw interfaces. In addition, I'm no longer convinced that they should cross-implement each other's base interface because it causes overload ambiguity.

**Comment by stealthcode** [comment]
> I'm no longer convinced that they should cross-implement

I think I agree with that statement. Do you mean that `Completable` should not implement the `ConsumableObservable`? I did not intend that in my POC.

**Comment by akarnokd** [comment]
`Flowable` looks like it could do both `ConsumableObservable` and `Publisher` or `Singe` could also implement `Publisher`, but some methods become ambiguous.

**Comment by stealthcode** [comment]
Yeah, since `Flowable`'s subscribe takes a `Subscriber` and both `ConsumableSubscriber` and `Publisher` both agree on the subscriber type then we should make Flowable implement both Consumable and Publisher. The others are difficult.

**Comment by akarnokd** [comment]
Added `ConsumableFlowable`. I had trouble with typing extend, using `Function<? super ConsumableFlowable<T>, R>` is too restrictive and gives no access to `Flowable's methods` and having `U extends ConsumableFlowable<T>>, Function<? super U, R>` has a type-bounce of ? super ? extends ConsumableFlowable.

**Comment by stealthcode** [comment]
I got extend to work with this.

```java
public class Flowable<T> implements Consumable<Subscriber<? super T>> {
@Override
    public <S2, X extends Consumable<S2>> X extend(Function<Consumer<Subscriber<? super T>>, X> convertion) {
        return convertion.apply(onSubscribe);
    }
}
```

So you could probably implement it similarly.

**Comment by stealthcode** [comment]
I'll try it out locally and see what works (and what doesn't).

**Comment by stealthcode** [comment]
I got it to work but the `Publisher<T> onSubscribe` is not a `Consumer<Subscriber<? super T>` so it'd require a wrapper.

```java
@Override
public <S2, X extends Consumable<S2>> X extend(Function<Consumer<Subscriber<? super T>>, X> conversion) {
    return conversion.apply(new Consumer<Subscriber<? super T>>() {
        @Override
        public void accept(Subscriber<? super T> subscriber) {
            onSubscribe.subscribe(subscriber);
        }});
}
```

This required a base interface for `ConsumableFlowable`. 

```java
public interface Consumable<S> {
    void subscribe(S subscriber);
    public <S2, X extends Consumable<S2>> X extend(Function<Consumer<S>, X> conversion);
}

public interface ConsumableFlowable<T> extends Consumable<Subscriber<? super T>> {
}

public class Flowable<T> implements Publisher<T>, ConsumableFlowable<T> {
    // ...
}
```

**Comment by akarnokd** [comment]
Let's experiment with that in a separate PR; also, I'd wait until the Flowable isn't a wrapper anymore but a base class that gets extended by all internal operators that were create/lift before.

**Comment by stealthcode** [comment]
Why do we need a separate PR? Extend is part one 1.x and with the
Consumable it's even more necessary to switch between types. Sure you can
create custom instance operators for converting between them or subscribe
but this interface is explicitly for converting between and chaining
together different streams. So it's useful to convert between types that do
not necessarily have a direct relationship (I.e. in 2 unrelated library
jars) So I am very disinclined to leave it out.

**Comment by akarnokd** [comment]
Why do you want this single PR to host every modification? There is a lot to do with 2.x and some of them, like extend, can be incremental.

**Comment by stealthcode** [comment]
I can see that it could be implemented incrementally. This PR didn't say
anything to that effect. I think that as a team the RxJava committers
should come to discuss and agree on the 2.0 Design doc. This process should
or may include additional elaboration or amendments to areas that do not
yet have team commitment. This would include items like extend (which you
seem to be uncertain in your support of). And in that process team members
would commit to and merge their  commitments incrementally in the design
doc on the 2.x branch before PRs of the uncommitted features are opened.

If any of us have concerns, suggestions, or alternatives or require
elaboration on the design doc then we should comment or open an issue/pr to
the design. I think we should wait until the Consumable and/or extend are
committed and merged to 2.x in the design doc so others have had a chance
to review and commit to it before continuing our work on this PR.

File: src/main/java/io/reactivex/ConsumableFlowable.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+
+package io.reactivex;
+
+import org.reactivestreams.Subscriber;
+
+/**
+ * A ConsumableFlowable represents a backpressured sequence of values optionally followed
+ * by a teraminal signal.
```
Comment (runningcode) [inline src/main/java/io/reactivex/ConsumableFlowable.java]: teraminal -> terminal

**Comment by hzsweers** [comment]
@stealthcode the design doc PR is merged, is there anything preventing this PR from proceeding?

**Comment by stealthcode** [comment]
@hzsweers Nope. Apologies for the delay but I won't be getting back to this until next week. Thanks for the reminder.

**Comment by ZacSweers** [comment]
@stealthcode ping

**Comment by akarnokd** [comment]
I'll redo this PR in multiple phases.

**Pull request closed without merging**